xmtp_api_grpc/src/streams/non_blocking_request.rs

```diff
@@ -0,0 +1,99 @@
+//! Non Blocking Request
+//! Fully awaits server_streaming request on native
+//! polls server_streaming request once on wasm (requiring Unpin)
+use crate::streams::NonBlockingWebStream;
+use futures::FutureExt;
+use pin_project_lite::pin_project;
+use prost::bytes::Bytes;
+use std::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tonic::{Response, Status};
+
+pin_project! {
+    pub struct NonBlockingStreamRequest<F> {
+        #[pin] inner: F,
+    }
+}
+
+impl<F> NonBlockingStreamRequest<F> {
+    pub fn new(inner: F) -> Self {
+        Self { inner }
+    }
+}
+
+impl<F> Future for NonBlockingStreamRequest<F>
+where
+    F: Future,
+{
+    type Output = F::Output;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.as_mut().project();
+        this.inner.poll(cx)
+    }
+}
+
+impl<F> NonBlockingStreamRequest<F>
+where
+    F: Future + Unpin,
+{
+    pub fn establish(&mut self) {
+        // we need to poll the future once to progress the future state &
+        // actually send the request in the first place.
+        // since the request has not even been sent yet, this should always be pending.
+        let noop_waker = futures::task::noop_waker();
+        let mut cx = std::task::Context::from_waker(&noop_waker);
+        {
+            let mut this = Pin::new(self);
+            if this.poll_unpin(&mut cx).is_ready() {
+                tracing::error!("Stream ready before established");
+                unreachable!("Stream ready before established")
+            }
+        }
+    }
```

Comment:
there's [poll_immediate](https://docs.rs/futures/latest/futures/future/fn.poll_immediate.html) in the futures crate that does this. But it correctly registers re-wakes because it is using the same context as the future will be polled with. Probably need to change the function to be async and await it, but it guarantees to return immediately and I think would be more clear what is going on with it being async as well.

Reply:
fixed! didn't know poll_immediate existed, definitely preferable to the noop waker